src/marshmallow/schema.py

```diff
@@ -584,7 +585,7 @@ def _deserialize(
         partial=None,
         unknown=RAISE,
         index=None,
-    ) -> _T | list[_T]:
+    ) -> dict | list[dict]:
```

Comment:
related to above: `_deserialize` is expected to return an arbitrary object or list of objects, not dictionaries.

Reply:
`ret_d = self.dict_class() ... return ret_d` implies that the return type of `_deserialize()` is the return type of `dict_class()` (`Mapping`). That change reveals that our usage of the return value actually depends on the `dict` interface. The fact that the type checker doesn't error on key access to a `TypeVar` which is effectively `Any`, seems like a mypy bug, but maybe it just needs explicit test coverage with a concrete type.

Enveloping occurs in `_do_load()`, so the return type of `_deserialize()` should not constrained by a `TypeVar`. Now that I understand the purpose of the `TypeVar` it looks like it should be moved to `load()` rather than removed.